controller/dataplane/owned_deployment.go

```diff
@@ -306,17 +318,55 @@ func reconcileDataPlaneDeployment(
 
 		k8sresources.SetDefaultsPodTemplateSpec(&desired.Spec.Template)
 
+		// Keep track of this for logging purposes
+		originalReplicaCount := existing.Spec.Replicas
+
 		// ensure that object metadata is up to date
 		updated, existing.ObjectMeta = k8sutils.EnsureObjectMetaIsUpdated(existing.ObjectMeta, desired.ObjectMeta)
 
-		// some custom comparison rules are needed for some PodTemplateSpec sub-attributes, in particular
-		// resources and affinity.
+		// Check if this is a deployment restart (kubectl rollout restart)
+		// We need to check if the existing deployment has the restart annotation
+		// This is different from our previous approach because the desired deployment
+		// is generated from the DataPlane resource and won't have the restart annotation
+		isRestartOperation := false
+
+		// Check if we have a fresh restart annotation (from the last 5 minutes)
+		// This prevents us from treating all deployments with old restart annotations
+		// as perpetually in a restart state, fixing issue #1390
+		hasRestartAnnotation, restartTimeStr := podTemplateSpecHasRestartAnnotation(&existing.Spec.Template)
+		if hasRestartAnnotation {
+			// Only treat it as a restart if the timestamp is recent (within 5 minutes)
+			restartTime, err := time.Parse(time.RFC3339, restartTimeStr)
+			if err == nil {
+				// Check if restart annotation is less than 5 minutes old
+				if time.Since(restartTime) < 5*time.Minute {
+					isRestartOperation = true
+					log.Debug(logger, "detected recent restart operation", "isRestartOperation", isRestartOperation, "restartTime", restartTime)
+				} else {
+					log.Debug(logger, "found old restart annotation, not treating as restart", "restartTime", restartTime)
+				}
+			} else {
+				isRestartOperation = true // If we can't parse time, assume it's a restart for safety
+				log.Debug(logger, "detected restart with unparseable timestamp", "timestamp", restartTimeStr)
+			}
+		}
```

Comment:
Can we encapsulate this whole piece of logic in a separate function to prevent `reconcileDataPlaneDeployment` becoming any bigger and to make it more readable?

Reply:
Are you only referring to the selected part? Or does it include the overall logic related to restart or scale?

In fact, some logs here can be removed.

Reply:
The selected part. If possible I think we could also pass the desired spec, explaining  why we need it ( to set the restart annotation if found, to prevent changing this).